 # <div align="center">Sécurisation des communications</div>

Pour qu’un message entre deux machines dans un réseau informatique ne puisse pas être compréhensible s’il est intercepté, il faut qu’il soit chiffré.
Il y a deux manières principales pour le faire:
   * Par un chiffrement symétrique qui utilise une clé unique, connue seulement de l’émetteur et du récepteur. L’émetteur chiffre le message avec la clé et le récepteur le déchiffre avec la même clé.
   * Par un chiffrement asymétrique qui utilise un couple de clé, l’une publique connue de tout le monde, l’autre privée connue uniquement par le récepteur. L’émetteur chiffre le message avec la clé publique. Le récepteur le déchiffre avec la clé privée. Ces deux méthodes de chiffrement sont utilisées lorsqu’un navigateur demande une page web avec le protocole HTTPS.<br>


Dans le schéma ci-dessous, on a représenté les échanges de communications entre un client
et un serveur avec le protocole HTTPS.

![Crypto](img/https.png)
Les échanges entre client et serveur se font avec un chiffrement symétrique (clé K2) après que celle-ci ait été échangée avec un chiffrement asymétrique.

### <div align="center"> Le chiffrement symétrique </div>
Dans un **chiffrement symétrique**, la clé utilisée par l’expéditeur pour chiffrer le message est la même que celle utilisée par le récepteur pour déchiffrer le message.<br>

*Un premier exemple* : **Le codage de César** avec une clé de 13.
En ne considérant que les lettres en majuscule et non accentuées, notre alphabet se réduit à 26 lettres ABCDEFGHIJKLMNOPQRSTUVWXYZ)
Avec un décalage de 13 : (A!N) et (N!A). On utilise bien la même clé, il s’agit donc d’un chiffrement symétrique.<br>


Par exemple le message **'BONJOUR, COMMENT ALLEZ-VOUS'** se code par :<br>
**'OBAWBHE, PBZZRAG NYYRM-IBHF'** (seules les lettres ont été codées...) et le déchiffrement nous redonne bien le message d’origine.

A vous:
Écrire un programme en python qui réalise ce chiffrement de César avec une clé de 13. On pourra utiliser ce dictionnaire et la méthode get associée par exemple.

In [5]:
d = {chr(i+65): chr((i+13)%26+65) for i in range(26)}

Ce type de syntaxe ne vous semble pas familier ? essayer le code ci-dessous .

In [4]:
for i in range(26):
    print(chr(i+65) + " -> " +  chr((i+13)%26+65))

A -> N
B -> O
C -> P
D -> Q
E -> R
F -> S
G -> T
H -> U
I -> V
J -> W
K -> X
L -> Y
M -> Z
N -> A
O -> B
P -> C
Q -> D
R -> E
S -> F
T -> G
U -> H
V -> I
W -> J
X -> K
Y -> L
Z -> M


*Un second exemple* : Utilisation de **XOR**<br>
XOR est une opération de logique bit à bit qui renvoie 0 si les deux bits sont égaux et 1 sinon.
Une particularité de cette opération est : 
* Si b1 xor b2 = b3 
* alors b1 xor b3 = b2 
* et b2 xor b3 = b1


Ce qui permet un chiffrement symétrique :<br>
Prenons le mot **'bonjour'** dont le code Unicode est donné par ce programme


In [33]:
m = "bonjour"
for c in m:
    print(ord(c),end=' ')

98 111 110 106 111 117 114 

Soit : 98 111 110 106 111 117 114<br>
Le mot **'nsi'** va nous servir de clé : ('nsi' ! 110 115 105)<br>
La méthode consiste alors à aligner le mot et la clé en la répétant autant de fois que nécessaire et d’effectuer un XOR entre les codes Unicode:

| |  |  |  |  |  |  |
|---|---|---|---|---|---|---|
|b| o| n| j| o| u| r|
|98| 111 |110| 106| 111| 117| 114|
|n| s| i| n| s |i |n
|110| 115| 105| 110| 115| 105| 110|

En pratiquant un XOR lettre par lettre entre les nombres obtenus (après une écriture en binaire...), on obtient les codes Unicode suivant:<br>
12 28 7 4 28 28 28<br>
(que l’on pourrait transformer en caractères mais ce n’est pas très utile...)

Pour décoder le message, il suffit alors de recommencer l’opération avec les codes Unicode du message chiffré en utilisant la même clé:<br>
(12 xor 110 ! 98 donc ’b’) etc.<br>
En Python l’opérateur **^** permet d’effectuer un xor directement sur deux entiers.<br>
Le programme suivant chiffre un message en utilisant cette méthode:

In [14]:
def chiffre(message,key):
    c = []
    n = len(message)
    m = len(key)
    j = 0
    for i in range(n):
        c.append(ord(message[i]) ^ ord(key[j]))
        j = (j+1)%m
    return c
print(chiffre("bonjour","nsi"))

[12, 28, 7, 4, 28, 28, 28]


Ce qui donne en console:<br>

[47, 22, 29, 30, 135, 7, 23, 65, 89, 16, 27, 133, 31, 0, 3, 13, 83, 21,132, 30, 0, 23, 84, 5, 27, 157, 1, 69, 82] <br>

Chaque nombre peut être converti en caractère (le contenu complet du message est chiffré,y compris les caractères de ponctuation et d’espace).

**A vous:**<br>
* 1) Modifier la fonction chiffre pour qu’elle renvoie le message chiffré (avec les caractères)
* 2) Écrire une fonction dechiffre qui prend en paramètres le message chiffré et la clé et qui renvoie le message déchiffré

### <div align="center"> Chiffrement asymétrique </div>

Pour reprendre l’exemple du code de César, il suffit de choisir une autre clé que 13 pour obtenir un chiffrement asymétrique, c’est à dire où la clé de chiffrement et de déchiffrement ne sont pas identiques.
Si la clé de chiffrement est 9, la clé de déchiffrement est de 17 (26 - 9).

A vous:<br>
Écrire un programme qui chiffre et déchiffre le message 'VIVE NSI A JEAN ROSTAND' pour
illustrer le chiffrement asymétrique avec le code de César

### <div align="center"> Le système RSA simplifié</div>

**Le principe:** <br>
Prenons la lettre 'b' dont le code Unicode est 98<br>

**Methode:** <br>
La **clé publique** pour notre exemple est : (e;n) = (41;437)<br>
n = 437 est le produit de deux nombres premiers (437 = 23*19), on considère alors l’entier;<br>
f = (23 - 1) * (19 - 1) = 22 * 18 = 396 et on choisit e = 41 un nombre premier inférieur à f .<br>


Pour coder la lettre **'b'** on détermine le reste de la division Euclidienne de **98**<sup>e</sup> par n.
Ce qui donne ici **110**, c’est à dire la lettre **'n'**

In [32]:
(98**41)%437

110

In [34]:
chr((98**41)%437)

'n'

Pour **décoder** la lettre 'n' on utilise la clé privée qui est (d;n) = (29;437) et on calcule le reste de la division Euclidienne de 110<sup>d</sup> par n.

La **clé privée** pour notre exemple est : (d;n) = (29;437)<br>
où d est un entier tel que le reste de la division de e * d = 41 * 29 par f = 396 est 1.<br>
*Des méthodes en mathématiques permettent de déterminer d.*

Le reste de la division euclidienne de 110<sup>29</sup> par 437 est 98 qui donne la lettre 'b'.

In [23]:
(110**29)%437

98

In [35]:
chr((110**29)%437)

'b'

A vous :<br>
Écrire un programme qui chiffre et déchiffre un message en utilisant ce système avec les clés précédentes:
Voici d’autres clés: <br>
* clé publique (103 , 697) ; 
* clé privée (87 , 697)    